juju/origin.py

```diff
@@ -0,0 +1,195 @@
+
+        if track is None:
+            track = ""
+        self.track = track
+        self.risk = risk
+
+    @staticmethod
+    def parse(s):
+        """parse a channel from a given string.
+        Parse does not take into account branches.
+
+        """
+        if not s:
+            raise JujuError("channel cannot be empty")
+
+        p = s.split("/")
+
+        print(len(p))
+
+        risk = None
+        track = None
+        if len(p) == 1:
+            if Risk.valid(p[0]):
+                risk = p[0]
+            else:
+                track = p[0]
+                risk = str(Risk.STABLE)
+        elif len(p) == 2:
+            track = p[0]
+            risk = p[1]
+        else:
+            raise JujuError("channel is malformed and has too many components {}".format(s))
+
+        if risk is not None and not Risk.valid(risk):
+            raise JujuError("risk in channel {} is not valid".format(s))
+        if track is not None and track == "":
+            raise JujuError("track in channel {} is not valid".format(s))
+
+        return Channel(track, risk)
+
+    def normalize(self):
+        track = self.track if self.track != "latest" else ""
+        risk = self.risk if self.risk != "" else ""
+        return Channel(track, risk)
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            return self.track == other.track and self.risk == other.risk
+        return False
+
+    def __str__(self):
+        path = self.risk
+        if self.track != "":
+            path = "{}/{}".format(self.track, path)
+        return path
+
+
+class Platform:
+    """ParsePlatform parses a string representing a store platform.
+    Serialized version of platform can be expected to conform to the following:
+
+     1. Architecture is mandatory.
+     2. OS is optional and can be dropped. Series is mandatory if OS wants
+     to be displayed.
+     3. Series is also optional.
+
+    To indicate something is missing `unknown` can be used in place.
+
+    Examples:
+
+     1. `<arch>/<os>/<series>`
+     2. `<arch>`
+     3. `<arch>/<series>`
+     4. `<arch>/unknown/<series>`
+
+    """
+    def __init__(self, arch, series=None, os=None):
+        self.arch = arch
+        self.series = series
+        self.os = os
+
+    @staticmethod
+    def parse(s):
+        if not s:
+            raise JujuError("platform cannot be empty")
+
+        p = s.split("/")
+
+        arch = None
+        os = None
+        series = None
+        if len(p) == 1:
+            arch = p[0]
+        elif len(p) == 2:
+            arch = p[0]
+            series = p[1]
+        elif len(p) == 3:
+            arch = p[0]
+            os = p[1]
+            series = p[2]
+        else:
+            raise JujuError("platform is malformed and has too many components {}".format(s))
+
+        if arch is None or arch == "":
+            raise JujuError("architecture in platform {} is not valid".format(s))
```

Comment:
I'd simplify this to:

```
if not arch:  # always required
    raise ...
```

That clearly calls the logic out as different from the two checks below, and makes it a little bit easier to parse for me.